File: packages/client-ts/src/events/subjects.ts
/**
 * SGRS NATS subject hierarchy.
 *
 * Structure:  sgrs.{category}.{tenant}.{id}.{event}
 *
 * NATS wildcard operators:
 *   *  — exactly one token    e.g. sgrs.scope.acme.*.finality.final
 *   >  — one or more tokens   e.g. sgrs.scope.acme.>
 *
 * Tenant isolation: subjects are namespaced per tenant.
 * For hard protocol-level isolation use one NATS Account per tenant.
 */

import { assertNatsSubjectSlug, TenantId } from "@sgrs/api-schema";

export const SGRS_PREFIX = "sgrs";

// ─── Token validation (same lexical rules as TenantId / scope slugs) ──────────

/**
 * Validate a NATS subject token — lowercase kebab-case slug, max length 120.
 *
 * @throws {Error} on empty/whitespace or invalid characters
 */
export function tok(value: string): string {
  return assertNatsSubjectSlug(value);
}

/**
 * Assert that `subject` starts with the expected tenant prefix.
 * Used internally to prevent cross-tenant publish.
 *
 * @throws {Error} if subject is outside the tenant's namespace
 */
export function assertOwnedByTenant(subject: string, tenant: string): void {
  // sgrs.{category}.{tenant}.{rest}
  // Parts[0]=sgrs  [1]=category  [2]=tenantToken
  const parts = subject.split(".");
  const tenantToken = tok(tenant);
  if (parts.length < 3 || parts[2] !== tenantToken) {
    throw new Error(
      `[SECURITY] Subject "${subject}" does not belong to tenant "${tenant}". ` +
        `Cross-tenant publish rejected.`,
    );
  }
}

// ─── Subject builders ─────────────────────────────────────────────────────────

export const subjects = {
  scope: {
    created:           (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.created`,
    updated:           (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.updated`,
    deleted:           (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.deleted`,
    finalityChanged:   (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.finality.changed`,
    finalityNearFinal: (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.finality.near-final`,
    /** Terminal — triggers certificate issuance */
    finalityFinal:     (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.finality.final`,
    /** CRITICAL PATH — must propagate to all swarm agents */
    vetoActivated:     (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.veto.activated`,
    vetoLifted:        (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.veto.lifted`,

    // Governance domain event subjects
    claimAdded:            (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.claim.added`,
    driftDetected:         (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.drift.detected`,
    contradictionDetected: (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.contradiction.detected`,
    contradictionResolved: (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.contradiction.resolved`,
    riskIdentified:        (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.risk.identified`,
    documentIndexed:       (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.document.indexed`,
    epochCompleted:        (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.epoch.completed`,

    // Subscription wildcards
    allFor:            (t: string, s: string) => `${SGRS_PREFIX}.scope.${tok(t)}.${tok(s)}.>`,
    allScopes:         (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.>`,
    allFinality:       (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.finality.>`,
    allVetos:          (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.veto.>`,
    allClaims:         (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.claim.>`,
    allDrifts:         (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.drift.>`,
    allContradictions: (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.contradiction.>`,
    allRisks:          (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.risk.>`,
    allDocuments:      (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.document.>`,
    allEpochs:         (t: string)            => `${SGRS_PREFIX}.scope.${tok(t)}.*.epoch.>`,
  },

  model: {
    connected: (t: string, h: string) => `${SGRS_PREFIX}.model.${tok(t)}.${tok(h)}.connected`,
    revoked:   (t: string, h: string) => `${SGRS_PREFIX}.model.${tok(t)}.${tok(h)}.revoked`,
    allModels: (t: string)            => `${SGRS_PREFIX}.model.${tok(t)}.>`,
  },

  agent: {
    heartbeat:     (t: string, a: string) => `${SGRS_PREFIX}.agent.${tok(t)}.${tok(a)}.heartbeat`,
    taskStarted:   (t: string, a: string) => `${SGRS_PREFIX}.agent.${tok(t)}.${tok(a)}.task.started`,
    taskCompleted: (t: string, a: string) => `${SGRS_PREFIX}.agent.${tok(t)}.${tok(a)}.task.completed`,
    taskFailed:    (t: string, a: string) => `${SGRS_PREFIX}.agent.${tok(t)}.${tok(a)}.task.failed`,
    /** Queue-group — NATS delivers each task to exactly ONE worker in the group */
    queue:         (t: string, taskType: string) => `${SGRS_PREFIX}.agent.queue.${tok(t)}.${tok(taskType)}`,
    allAgents:     (t: string)            => `${SGRS_PREFIX}.agent.${tok(t)}.>`,
  },

  audit: {
    all:    (t: string) => `${SGRS_PREFIX}.audit.${tok(t)}.>`,
    scopes: (t: string) => `${SGRS_PREFIX}.audit.${tok(t)}.scope.>`,
    models: (t: string) => `${SGRS_PREFIX}.audit.${tok(t)}.model.>`,
    agents: (t: string) => `${SGRS_PREFIX}.audit.${tok(t)}.agent.>`,
  },
} as const;

/** All event subjects for a tenant (scope + model + agent).
 *  Returns a fixed-length tuple — destructure safely without `!` assertions.
 */
export function allTenantSubjects(tenant: string): [string, string, string] {
  return [
    subjects.scope.allScopes(tenant),
    subjects.model.allModels(tenant),
    subjects.agent.allAgents(tenant),
  ];
}

// ─── Stream names ─────────────────────────────────────────────────────────────

function streamKey(tenant: string): string {
  if (!tenant || !tenant.trim()) {
    throw new Error(
      `Cannot derive stream name from empty or whitespace-only tenant: ${JSON.stringify(tenant)}`,
    );
  }
  const t = tenant.trim();
  const parsed = TenantId.safeParse(t);
  if (!parsed.success) {
    throw new Error(`Invalid tenant for stream name: ${JSON.stringify(tenant)}`);
  }
  return parsed.data.toUpperCase().replace(/[^A-Z0-9]/g, "_");
}

/** JetStream stream name for a tenant's full audit log (7-year retention) */
export function auditStreamName(tenant: string): string {
  const key = streamKey(tenant); // throws on empty/whitespace
  if (!key) throw new Error(`Cannot derive audit stream name from empty tenant: ${JSON.stringify(tenant)}`);
  return `SGRS_AUDIT_${key}`;
}

/** JetStream stream name for a tenant's scope events */
export function scopeStreamName(tenant: string): string {
  const key = streamKey(tenant); // throws on empty/whitespace
  if (!key) throw new Error(`Cannot derive scope stream name from empty tenant: ${JSON.stringify(tenant)}`);
  return `SGRS_SCOPE_${key}`;
}

/**
 * Sanitise a value for use as a JetStream durable consumer name.
 * @throws {Error} on empty or whitespace-only input
 */
export function sanitiseDurable(value: string): string {
  if (!value || !value.trim()) {
    throw new Error(
      `JetStream durable consumer name must not be empty (got: ${JSON.stringify(value)})`,
    );
  }
  return value.replace(/[^A-Za-z0-9\-_]/g, "_").slice(0, 128);
}

File: packages/client-py/src/sgrs_client/events/subjects.py
"""SGRS NATS subject hierarchy.

Structure:  sgrs.{category}.{tenant}.{id}.{event}

Lexical rules for every subject token mirror ``TenantId`` / scope slugs from the
REST API schema: lowercase ``[a-z0-9]+`` optionally separated by single hyphens
(no underscores, uppercase, dots, shell wildcards, or leading/trailing hyphen).

Security: :func:`tok` validates rather than “best-effort” sanitizing, so malformed
identifiers fail fast instead of collapsing into ambiguous NATS namespaces.
"""

from __future__ import annotations

import re

SGRS_PREFIX = "sgrs"

_SLUG_LOWERCASE = re.compile(r"^(?:[a-z0-9]|[a-z0-9][a-z0-9-]*[a-z0-9])$")
_MAX_NATSTOKEN_LEN = 120
_MAX_TENANT_SLUG_LEN = 64

_TOK_SAN_RE = re.compile(r"[^A-Za-z0-9\-_]")  # durable consumer sanitisation


def tok(value: str) -> str:
    """Validate *value* for use as one dot-separated NATS subject token.

    Raises:
        ValueError: on empty/invalid slug or overly long tokens.
    """
    t = value.strip()
    if not t:
        raise ValueError(
            f"NATS subject token must not be empty (got: {value!r})",
        )
    if len(t) > _MAX_NATSTOKEN_LEN:
        raise ValueError(
            f"NATS subject token too long (max {_MAX_NATSTOKEN_LEN} chars): "
            f"{value[:24]!r}…",
        )
    if not _SLUG_LOWERCASE.fullmatch(t):
        raise ValueError(
            "Invalid NATS subject token: use lowercase a-z, digits, and hyphens "
            f"(no underscores or uppercase): {value!r}",
        )
    return t


def assert_owned_by_tenant(subject: str, tenant: str) -> None:
    """Assert that *subject* is within *tenant*'s namespace.

    Raises:
        ValueError: if subject belongs to a different tenant.
    """
    parts = subject.split(".")
    tenant_token = tok(tenant)
    if len(parts) < 3 or parts[2] != tenant_token:
        raise ValueError(
            f"[SECURITY] Subject {subject!r} does not belong to tenant "
            f"{tenant!r}. Cross-tenant publish rejected.",
        )


# ─── Subject builders ─────────────────────────────────────────────────────────


class _ScopeSubjects:
    def created(self, tenant: str, scope_id: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.created"

    def updated(self, tenant: str, scope_id: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.updated"

    def deleted(self, tenant: str, scope_id: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.deleted"

    def finality_changed(self, tenant: str, scope_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.finality.changed"
        )

    def finality_near_final(self, tenant: str, scope_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.finality.near-final"
        )

    def finality_final(self, tenant: str, scope_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.finality.final"
        )

    def veto_activated(self, tenant: str, scope_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.veto.activated"
        )

    def veto_lifted(self, tenant: str, scope_id: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.veto.lifted"

    # Subscription wildcards

    def all_for(self, tenant: str, scope_id: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.{tok(scope_id)}.>"

    def all_scopes(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.>"

    def all_finality(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.*.finality.>"

    def all_vetos(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.scope.{tok(tenant)}.*.veto.>"


class _ModelSubjects:
    def connected(self, tenant: str, handle: str) -> str:
        return (
            f"{SGRS_PREFIX}.model.{tok(tenant)}.{tok(handle)}.connected"
        )

    def revoked(self, tenant: str, handle: str) -> str:
        return f"{SGRS_PREFIX}.model.{tok(tenant)}.{tok(handle)}.revoked"

    def all_models(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.model.{tok(tenant)}.>"


class _AgentSubjects:
    def heartbeat(self, tenant: str, agent_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.agent.{tok(tenant)}.{tok(agent_id)}.heartbeat"
        )

    def task_started(self, tenant: str, agent_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.agent.{tok(tenant)}.{tok(agent_id)}.task.started"
        )

    def task_completed(self, tenant: str, agent_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.agent.{tok(tenant)}.{tok(agent_id)}.task.completed"
        )

    def task_failed(self, tenant: str, agent_id: str) -> str:
        return (
            f"{SGRS_PREFIX}.agent.{tok(tenant)}.{tok(agent_id)}.task.failed"
        )

    def queue(self, tenant: str, task_type: str) -> str:
        return (
            f"{SGRS_PREFIX}.agent.queue.{tok(tenant)}.{tok(task_type)}"
        )

    def all_agents(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.agent.{tok(tenant)}.>"


class _AuditSubjects:
    def all(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.audit.{tok(tenant)}.>"

    def scopes(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.audit.{tok(tenant)}.scope.>"

    def models(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.audit.{tok(tenant)}.model.>"

    def agents(self, tenant: str) -> str:
        return f"{SGRS_PREFIX}.audit.{tok(tenant)}.agent.>"


class _Subjects:
    scope = _ScopeSubjects()
    model = _ModelSubjects()
    agent = _AgentSubjects()
    audit = _AuditSubjects()


subjects = _Subjects()


def all_tenant_subjects(tenant: str) -> list[str]:
    return [
        subjects.scope.all_scopes(tenant),
        subjects.model.all_models(tenant),
        subjects.agent.all_agents(tenant),
    ]


def _stream_key(tenant: str) -> str:
    t = tenant.strip()
    if not t:
        raise ValueError(f"Cannot derive stream name from empty tenant: {tenant!r}")
    if len(t) > _MAX_TENANT_SLUG_LEN:
        raise ValueError(f"Invalid tenant for stream name: {tenant!r}")
    if not _SLUG_LOWERCASE.fullmatch(t):
        raise ValueError(f"Invalid tenant for stream name: {tenant!r}")
    return re.sub(r"[^A-Z0-9]", "_", t.upper())


def audit_stream_name(tenant: str) -> str:
    return f"SGRS_AUDIT_{_stream_key(tenant)}"


def scope_stream_name(tenant: str) -> str:
    return f"SGRS_SCOPE_{_stream_key(tenant)}"


def sanitise_durable(value: str) -> str:
    if not value or not value.strip():
        raise ValueError(
            f"JetStream durable consumer name must not be empty (got: {value!r})",
        )
    return _TOK_SAN_RE.sub("_", value)[:128]
